### PR TITLE
Inject keycloak user token to bayesian LS

### PR DESCRIPTION
**Pull request by sunix**
- Keeping the last token used (from Keycloak)AuthenticationFilter) and make it available to the bayesian LS at start time.
- update `KeycloakAuthenticationFilter` so it works well with new upstream websocket URIs

### CODE REVIEW

File: builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java [hunk 1]
```diff
@@ -41,7 +42,10 @@
     private boolean keycloakDisabled;
 
     @Inject
-    private KeycloakUserChecker userChecker;
+    private KeycloakTokenStore         kcTokenStore;
```
Comment (tsmaeder) [inline builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java]: Field injection requires the use of some injection framework when writing tests. Personally, I'm not a fan. Why not use constructor injection? Is there an opinion on the Che team?

Comment (sunix) [inline builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java]: Personally, I find it more readable than useless constructor and parameters in constructor when it's actually doing nothing else then just setting the value to the field. Using injection or mock framework is not a problem to me.

**Comment by sunix** [comment]
I've added https://github.com/redhat-developer/rh-che/pull/237/commits/45acdcfb5668c9b217121974ca9e905b3baeb0d9 so it works well with the latest master

File: builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java [hunk 2]
```diff
@@ -128,10 +135,14 @@ private boolean isWorkspaceAgentRequest(String authHeader) {
...
-                || requestScheme.equals("wss") || requestURI.contains("/websocket/")
-                || requestURI.endsWith("/token/user");
+    private boolean isWsAgentUserTokenRequest(String requestURI) {
+        return requestURI.endsWith("/token/user");
+    }
+
+    private boolean isWebsocketRequest(String requestURI, String requestScheme, String upgradeHeader) {
+        return "websocket".equals(upgradeHeader) && //
+               (requestURI.endsWith("/ws") || requestURI.endsWith("/eventbus") || requestScheme.equals("ws") || requestScheme.equals("wss")
+                || requestURI.contains("/websocket") || requestURI.contains("/wsagent"));
```
Comment (davidfestal) [inline builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java]: it seems that, if you explicitly add the `websocket` upgrade header (in a curl request for example), you can access to many  wsagent rest endpoints without any security.

Comment (sunix) [inline builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java]: there is a `&&`

Comment (davidfestal) [inline builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java]: sure, but the condition:
```
"websocket".equals(upgradeHeader) && requestURI.contains("/wsagent")
```
seem to allow many wsagent requests, due to the use of the `contains()` method.

Comment (sunix) [inline builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java]: ok so the problem is not about websocket upgrade header :)

Comment (sunix) [inline builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java]: fixed

File: builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java [hunk 3]
```diff
@@ -128,10 +135,14 @@ private boolean isWorkspaceAgentRequest(String authHeader) {
...
-    private boolean isWebsocketRequest(String requestURI, String requestScheme) {
-        return requestURI.endsWith("/ws") || requestURI.endsWith("/eventbus") || requestScheme.equals("ws")
-                || requestScheme.equals("wss") || requestURI.contains("/websocket/")
-                || requestURI.endsWith("/token/user");
+    private boolean isWsAgentUserTokenRequest(String requestURI) {
+        return requestURI.endsWith("/token/user");
+    }
+
+    private boolean isWebsocketRequest(String requestURI, String requestScheme, String upgradeHeader) {
+        return "websocket".equals(upgradeHeader) && //
```
Comment (ibuziuk) [inline builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java]: @sunix are you sure that all websocket requests **must** contain "Upgrade" header ?

Comment (ibuziuk) [inline builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java]: also, it is not clear for me why are you using // ?

File: builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java (hunk 2 above)
Comment (davidfestal) [inline builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java]: thanks !

File: builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java (hunk 3 above)
Comment (sunix) [inline builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java]: yes websocket is an upgrade of http
`//` is a trick to keep carriage return even if the formatter wouldn't do so
so it's to make the code more readable and avoid other people breaking the formatting chosen

**Pull request merged**